accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,170 @@
+
+This requires a mapping between the MSBuild property name, the full name of the feature switch for runtime configuration (which would show up in `.runtimeconfig.json`) and the read-only static property in the managed code which is used to branch the behavior. The mappings will be defined by [`RuntimeHostConfigurationOptions`](https://github.com/dotnet/sdk/blob/36ef8b2aa8e5d579c921704bdab69a7407936889/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L347) which includes an AppContext configuration name and its value based on the values of user-facing MSBuild properties, and by `substitutions.xml` files embedded in the assembly defining the features.
+
+*Note that it seems likely that some of switches may require 1:many mapping because they're targeting existing code which uses multiple properties to determine the presence of the feature. The `substitutions.xml` will allow 1:many mappings if needed.*
+
+Ultimately some of these switches may also need VS UI integration, but for now most of the feature switches should be perfectly fine with only MSBuild properties.
+
+The ability to specify these from MSBuild projects as properties also means that different templates and SDKs can set different defaults - so for example Blazor or Xamarin SDKs may choose to turn off some of these features by default.
+
+Example definition of a feature flag with SDK support in [`Microsoft.NET.Sdk.targets`](https://github.com/dotnet/sdk/blob/36ef8b2aa8e5d579c921704bdab69a7407936889/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets) or by SDK components:
+```xml
+<ItemGroup>
+    ...
+    <RuntimeHostConfigurationOption Include="System.Runtime.OptionalFeatureBehavior"
+                                    Condition="'$(OptionalUserFacingBehavior)' != ''"
+                                    Value="$(OptionalUserFacingBehavior)" />
+    ...
+</ItemGroup>
+```
+
+The name of the property should be picked so that it's clear what `true`/`false` mean. In this case, the idea is that `false` means the feature will be disabled. Other cases (for example `InvariantGlobalization`) might have the opposite polarity.
+
+### Generate the right input for the linker in SDK
+
+All names/values from `RuntimeHostConfigurationOptions` will be passed to the ILLink task, which will apply any feature substitutions defined in [`substitutions.xml`](https://github.com/mono/linker/blob/master/src/linker/README.md#using-custom-substitutions). The substitutions file format will be extended to condition the substitutions based on the feature name/value. Any `RuntimeHostConfigurationOptions` which do not have feature implementations in `substitutions.xml` will not result in any modifications to the IL.
+
+Example of a feature implementation:
+
+```xml
+<linker>
+  <assembly fullname="System.Runtime.FeatureDefiningAssembly" feature="System.Runtime.OptionalFeatureBehavior" featurevalue="false">
```

Comment:
I think this should better be feature="OptionalUserFacingBehavior". Secondly the feature should be on <linker> node not at <assembly> to allow grouping across assemblies

Reply:
Why do you think it should be based on the MSBuild property (`OptionalUserFacingBehavior`) and not on the runtime config property (`System.Runtime.OptionalFeatureBehavior`). Note that the implementation (C# of the library) should use `AppContext.TryGetSwitch("System.Runtime.OptionalFeatureBehavior")` (the runtime config property) in order to get consistent behavior between linked and non-linked versions of the app. And since the `substitution.xml` is very close to the actual implementation in the library I actually find it nice that they both talk about the same property.

In that sense the MSBuild is really only a syntax sugar for `csproj` (and potentially for VS UI one day).

Reply:
> Secondly the feature should be on node not at to allow grouping across assemblies

This I don't understand. Do you mean it should look something like:
```XML
<feature name="System.Runtime.OptionalFeatureBehavior" value="false">
  <assembly fullname="System.Runtime.FeatureDefiningAssembly">
    <type fullname="System.Runtime.FeatureDefiningType">
      <method signature="System.Boolean get_IsOptionalFeatureEnabled()" body="stub" value="false"/>
    </type>
  </assembly>
</feature>
```

Reply:
Sorry for not being clear. I thought about something like
```
<linker feature="System.Runtime.OptionalFeatureBehavior" 
featurevalue="false">
	...
</linker>
```

I haven't thought about allowing feature on every node which I think settles this down.

Reply:
The one case I'm struggling with is if we allow non-boolean feature values. Then the current approach doesn't work that well (since there's only one feature value defined per node). But I guess we can live with that....?

Reply:
If we want to support, say, integer substitutions where the integer can vary over a wide range (and assuming we don't want to put every possible value into the xml), we'd need to extend the format anyway, something like:
```xml
<method signature="System.Int32 get_FeatureInt()" body="$(FeatureInt)" />
```
where we substitute the `$(FeatureInt)` with a value known at link time. In this case the extra indirection means we will be substituting values into the substitution definitions which will be substituted into IL :)

The feature conditions just make this more redundant because it's likely we'll use a pattern like:
```xml
<assembly feature="IntegerFeature" featurevalue="$(FeatureInt)">
...
  <method signature="System.Int32 get_FeatureInt()" body="$(FeatureInt)" />
...
</assembly>
```
...or maybe in the inner scope we could reference the `featurevalue` of the outer scope... this quickly turns into a "let's design an XML-based programming language" which we already have enough of in our ecosystem. ;)

I'm hoping we can use the boolean features to get off the ground quickly, and later settle on some kind of attribute-based feature definitions instead.

Reply:
> I'm hoping we can use the boolean features to get off the ground quickly, and later settle on some kind of attribute-based feature definitions instead.

I agree. Let's not design another DSL, if there is a need to propagate the input value we could come up with something later

Reply:
Just asking ... perfectly OK to support just booleans.

Reply:
Sounds good. I'm glad you brought it up - it's definitely worth noting that it doesn't extend nicely to non-booleans.